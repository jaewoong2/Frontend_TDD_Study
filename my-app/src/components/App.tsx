import useChange from 'hooks/useChange';
import useLists from 'hooks/useLists';
import styled from 'styled-components';
import { Forms } from './blocks/Forms';
import Lists from './blocks/Lists';

function App() {
  const [todo, updateState, onChangeTodo] = useChange('');
  const { lists, addList, removeList } = useLists([]);

  return (
    <Container>
      <Contents>
        <Lists lists={lists} handleDelete={removeList} />
        <Forms
          removeTodo={() => updateState('')}
          handleAddTodo={addList(todo)}
          onChangeTodo={onChangeTodo}
          todo={todo}
          placeHolder="할 일을 입력해 주세요"
        />
      </Contents>
    </Container>
  );
}

export default App;

const Container = styled.div`
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Contents = styled.div`
  display: flex;
  flex-direction: column;
  background-color: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
`;
