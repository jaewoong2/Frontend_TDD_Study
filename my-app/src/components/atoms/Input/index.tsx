import { forwardRef } from 'react';
import styled from 'styled-components';

export const InputContainer = styled.label`
  display: flex;
`;

const StyledInput = styled.input`
  font-size: 16px;
  margin-right: 9px;
  padding: 10px 10px;
  border-radius: 8px;
  border: 1px solid #bdbdbd;
  outline: none;
  flex: 1;
`;

type Props = React.InputHTMLAttributes<HTMLInputElement>;

export const Input = forwardRef<HTMLInputElement, Props>(
  ({ value, onChange, placeholder }, ref) => {
    return (
      <StyledInput
        value={value}
        onChange={onChange}
        placeholder={placeholder}
        ref={ref}
      />
    );
  },
);

Input.displayName = 'Input';
Input.defaultProps = {
  onChange: () => {},
};
