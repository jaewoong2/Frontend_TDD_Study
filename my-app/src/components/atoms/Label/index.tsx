import styled from 'styled-components';

export const Label = styled.label`
  color: #fff;
  font-size: 16px;
`;
