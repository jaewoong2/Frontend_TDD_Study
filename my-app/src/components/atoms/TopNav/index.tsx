import React from 'react';
import styled from 'styled-components';

type Props = {
  children?: React.ReactNode;
  left?: React.ReactNode;
  right?: React.ReactNode;
};

function TopNav({ children, left, right }: Props) {
  return (
    <Nav>
      <div className="box left">{left}</div>
      <div className="box center">{children}</div>
      <div className="box right">{right}</div>
    </Nav>
  );
}

export { TopNav };

const Nav = styled.nav`
  width: 100%;
  height: 40px;
  background-color: #9bceee;
  display: flex;
  border-radius: 8px;
  justify-content: center;
  align-items: center;

  .box {
    width: 100%;
  }

  .center {
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;
