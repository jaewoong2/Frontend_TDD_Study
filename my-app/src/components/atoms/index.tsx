export { Button } from './Button';
export { Input, InputContainer } from './Input';
export { Item } from './Item';
