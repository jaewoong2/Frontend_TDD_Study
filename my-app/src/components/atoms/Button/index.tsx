import React from 'react';
import styled from 'styled-components';
import { Label } from '../Label';

type Props = {
  backgroundColor?: string;
  hoverColor?: string;
} & React.DetailsHTMLAttributes<HTMLButtonElement>;

export const Button = React.forwardRef<HTMLButtonElement, Props>(
  ({ backgroundColor, hoverColor, children, onClick }, ref) => {
    return (
      <Btn
        backgroundColor={backgroundColor}
        onClick={onClick}
        hoverColor={hoverColor}
        ref={ref}>
        <Label>{children}</Label>
      </Btn>
    );
  },
);

Button.displayName = 'Button';
Button.defaultProps = {
  backgroundColor: '#304FFE',
  hoverColor: '#1E40FF',
};

const Btn = styled.button<Props>`
  outline: none;
  border: none;
  text-align: center;
  background-color: ${({ backgroundColor }) => backgroundColor};
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
  &:hover {
    background-color: ${({ hoverColor }) => hoverColor};
  }
  &:active {
    box-shadow: inset 5px 5px 10px #00000033;
  }
`;
