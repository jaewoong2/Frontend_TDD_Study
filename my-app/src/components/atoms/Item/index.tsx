import styled from 'styled-components';
import React from 'react';
import { Label } from '../Label';
import { Button } from '../Button';

type Props = {
  children: React.ReactNode;
  onClick?: () => void;
  hoverColor?: string;
  backgroundColor?: string;
  buttonName?: string;
};

function Item({
  children,
  onClick,
  hoverColor,
  backgroundColor,
  buttonName,
}: Props) {
  return (
    <ItemContainer>
      <Label>{children}</Label>
      <Button
        onClick={onClick}
        hoverColor={hoverColor}
        backgroundColor={backgroundColor}>
        {buttonName}
      </Button>
    </ItemContainer>
  );
}

export { Item };

Item.defaultProps = {
  onClick: () => {
    console.log('onClick Button');
  },
  hoverColor: '',
  buttonName: '',
  backgroundColor: '',
};

const ItemContainer = styled.li`
  display: flex;
  align-items: center;
  border-bottom: 1px solid #bdbdbd;
  margin: 10px;
  padding: 10px;

  label {
    flex: 1;
    color: black;
  }
`;
