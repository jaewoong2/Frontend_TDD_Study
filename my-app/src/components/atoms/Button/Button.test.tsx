import { fireEvent, render, screen } from '@testing-library/react';
import { Button } from 'components/atoms';
import 'jest-styled-components';

describe('<Button />', () => {
  it('renders component correclty', () => {
    const { container } = render(<Button>Button Test</Button>);
    const label = screen.getByText('Button Test');
    expect(label).toBeInTheDocument();

    const parent = label.parentElement;
    expect(parent).toHaveStyleRule('background-color', '#304FFE');
    expect(parent).toHaveStyleRule('background-color', '#1E40FF', {
      modifier: ':hover',
    });

    expect(container).toMatchSnapshot();
  });

  it('changes backgroundColor and hoverColor Props', () => {
    const BACKGROUND_COLOR = '#FF1744';
    const HOVER_COLOR = '#F01440';
    render(
      <Button backgroundColor={BACKGROUND_COLOR} hoverColor={HOVER_COLOR}>
        Button Test
      </Button>,
    );

    const parent = screen.getByText('Button Test').parentElement;
    expect(parent).toHaveStyleRule('background-color', BACKGROUND_COLOR);
    expect(parent).toHaveStyleRule('background-color', HOVER_COLOR, {
      modifier: ':hover',
    });
  });

  it('clicks the button', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Button Test</Button>);

    const label = screen.getByText('Button Test');
    fireEvent.click(label);
    /**
     * error - Call Time 1 (Now)
     * expect(handleClick).toHaveBeenCalledTimes(0);
     */
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
