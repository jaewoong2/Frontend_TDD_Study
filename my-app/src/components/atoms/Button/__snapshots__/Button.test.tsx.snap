// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders component correclty 1`] = `
.c1 {
  color: #fff;
  font-size: 16px;
}

.c0 {
  outline: none;
  border: none;
  text-align: center;
  background-color: #304FFE;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c0:hover {
  background-color: #1E40FF;
}

.c0:active {
  box-shadow: inset 5px 5px 10px #00000033;
}

<div>
  <button
    class="c0"
  >
    <label
      class="c1"
    >
      Button Test
    </label>
  </button>
</div>
`;
